subsys/bluetooth/host/hci_core.c

```diff
@@ -2636,10 +2636,12 @@ int bt_unpair(u8_t id, const bt_addr_le_t *addr)
 		return -EINVAL;
 	}
 
+#if defined(CONFIG_BT_SMP) || defined(CONFIG_BT_BREDR)
```

Comment:
It looks like bt_foreach_bond is only defined if `CONFIG_BT_SMP`, so ` || defined(CONFIG_BT_BREDR)` should be removed.
Correct me if i'm wrong.

Reply:
Done.

Reply:
Is that correct though? Why don't we want `bt_foreach_bond()` to be available for BR/EDR? Seems like a perfectly valid use case for it, and `bt_unpair()` (which calls it) should work for BR/EDR as well.

Reply:
> Is that correct though? Why don't we want `bt_foreach_bond()` to be available for BR/EDR? Seems like a perfectly valid use case for it, and `bt_unpair()` (which calls it) should work for BR/EDR as well.

Yes, `bt_foreach_bond()` is also flexible for BR/EDR, but this is only in the case of SMP enabled, for BR/EDR, I think it is enough to call `unpair()` directly if the br link key is activated.

Reply:
I just realised BT_BREDR Kconfig option does `select BT_SMP` so you can't even have one without the other.

Reply:
It is expected to check the SMP without BREDR, the bond operation only make sense if SMP is enabled.

Reply:
I'm still struggling to understand this. BR/EDR also has bonds, and those are represented by the same bt_keys structure in Zephyr as LE bonds. That said, as per my previous finding *if* BR/EDR is enabled then checking for SMP will cover it, however I wonder if it would make sense to have some virtual Kconfig option that both SMP and BR/EDR select (e.g BT_KEYS). That would at least make this condition more intuitive. That said, it doesn't detract from the correctness of this PR, so I guess it can be done as a later fix.

Reply:
Agree with you, but from view of the current code dependency, run the `BREDR` without `SMP` needs some refine further.

Reply:
@jhedberg It looks like there is issues with BR/EDR and bonds to me.
There are keys_br as well, which has the bond information from SSP (as far as I understand, I don't know BR/EDR) which is not saved (open PR: https://github.com/zephyrproject-rtos/zephyr/pull/23044/files) and is not covered by the bt_unpair and bt_bonds_foreach functions.

This PR on it's own looks correct to me.

Reply:
Ok, fair enough. Thanks for clarifying.

Reply:
Well not anymore ...

Since bt_bonds_foreach is defined in keys.c and keys.c is only included if `CONFIG_BT_SMP` so I would remove the `|| CONFIG_BT_BREDR`, as it stands it is not correct, and BR/EDR will have to be fixed independently of this PR.